File: ios/Classes/vungle/VungleBannerAdManager.swift
import Flutter
import UIKit
import VungleAdsSDK

class VungleBannerAdManager: NSObject {
    private let listenerId: String
    private let placementId: String
    private let bannerSize: Int
    private var channel: FlutterMethodChannel
    private var bannerAd: VungleBanner?
    private var bannerView: UIView?
    
    init(listenerId: String, placementId: String, bannerSize: Int, channel: FlutterMethodChannel) {
        self.listenerId = listenerId
        self.placementId = placementId
        self.bannerSize = bannerSize
        self.channel = channel
        super.init()
    }
    
    private func getVungleBannerSize() -> BannerSize {
        switch bannerSize {
        case 0:
            return .regular // 320x50
        case 1:
            return .short // 300x50
        case 2:
            return .leaderboard // 728x90
        case 3:
            return .mrec // 300x250
        default:
            return .regular
        }
    }
    
    func loadAd() {
        print("[VungleBannerAd] Loading ad - listenerId: \(listenerId), placementId: \(placementId), size: \(bannerSize)")
        
        bannerAd = VungleBanner(placementId: placementId, size: getVungleBannerSize())
        bannerAd?.delegate = self
        bannerAd?.load()
    }
    
    func getBannerView() -> UIView? {
        return bannerView
    }
    
    func dispose() {
        print("[VungleBannerAd] Disposing ad - listenerId: \(listenerId)")
        bannerAd?.delegate = nil
        bannerAd = nil
        bannerView = nil
    }
    
    private func sendCallback(_ methodName: String, additionalData: [String: Any] = [:]) {
        var args: [String: Any] = [
            "listenerId": listenerId
        ]
        args.merge(additionalData) { (_, new) in new }
        channel.invokeMethod(methodName, arguments: args)
    }
}

// MARK: - VungleBannerDelegate

extension VungleBannerAdManager: VungleBannerDelegate {
    func bannerAdDidLoad(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdLoaded - listenerId: \(listenerId)")
        sendCallback("onBannerAdLoaded")
    }
    
    func bannerAdDidFailToLoad(_ banner: VungleBanner, withError: NSError) {
        print("[VungleBannerAd] onAdLoadFailed - listenerId: \(listenerId), error: \(withError.localizedDescription)")
        sendCallback("onBannerAdLoadFailed", additionalData: [
            "error": ["code": withError.code, "message": withError.localizedDescription]
        ])
    }
    
    func bannerAdWillPresent(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdWillPresent - listenerId: \(listenerId)")
    }
    
    func bannerAdDidPresent(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdShowed - listenerId: \(listenerId)")
        sendCallback("onBannerAdShowed")
    }
    
    func bannerAdDidFailToPresent(_ banner: VungleBanner, withError: NSError) {
        print("[VungleBannerAd] onAdFailedToPresent - listenerId: \(listenerId), error: \(withError.localizedDescription)")
        sendCallback("onBannerAdLoadFailed", additionalData: [
            "error": ["code": withError.code, "message": withError.localizedDescription]
        ])
    }
    
    func bannerAdDidTrackImpression(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdImpression - listenerId: \(listenerId)")
        sendCallback("onBannerAdImpression")
    }
    
    func bannerAdDidClick(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdClicked - listenerId: \(listenerId)")
        sendCallback("onBannerAdClicked")
    }
    
    func bannerAdWillLeaveApplication(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdWillLeaveApplication - listenerId: \(listenerId)")
    }
    
    func bannerAdWillClose(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdWillClose - listenerId: \(listenerId)")
    }
    
    func bannerAdDidClose(_ banner: VungleBanner) {
        print("[VungleBannerAd] onAdClosed - listenerId: \(listenerId)")
        sendCallback("onBannerAdClosed")
    }
}

File: android/src/main/kotlin/com/example/multi_ads/vungle/VungleBannerAdManager.kt
package com.example.multi_ads.vungle

import android.app.Activity
import android.content.Context
import android.util.Log
import android.view.View
import io.flutter.plugin.common.MethodChannel
import com.vungle.ads.BannerAdListener
import com.vungle.ads.BaseAd
import com.vungle.ads.VungleAdSize
import com.vungle.ads.VungleBannerView
import com.vungle.ads.VungleError

class VungleBannerAdManager(
    private val context: Context,
    private val activity: Activity?,
    private val listenerId: String,
    private val placementId: String,
    private val bannerSize: Int,
    private val channel: MethodChannel
) : BannerAdListener {

    companion object {
        const val TAG = "VungleBannerAd"
    }

    private var bannerView: VungleBannerView? = null

    private fun getVungleAdSize(): VungleAdSize {
        return when (bannerSize) {
            0 -> VungleAdSize.BANNER // 320x50
            1 -> VungleAdSize.BANNER_SHORT // 300x50
            2 -> VungleAdSize.BANNER_LEADERBOARD // 728x90
            3 -> VungleAdSize.MREC // 300x250
            else -> VungleAdSize.BANNER
        }
    }

    fun loadAd() {
        Log.d(TAG, "Loading banner ad - listenerId: $listenerId, placementId: $placementId, size: $bannerSize")
        
        bannerView = VungleBannerView(context, placementId, getVungleAdSize()).apply {
            adListener = this@VungleBannerAdManager
            load()
        }
    }

    fun getBannerView(): View? {
        return bannerView
    }

    fun dispose() {
        Log.d(TAG, "Disposing banner ad - listenerId: $listenerId")
        bannerView?.finishAd()
        bannerView?.adListener = null
        bannerView = null
    }

    private fun sendCallback(methodName: String, additionalData: Map<String, Any?> = emptyMap()) {
        activity?.runOnUiThread {
            val args = mutableMapOf<String, Any?>(
                "listenerId" to listenerId
            )
            args.putAll(additionalData)
            channel.invokeMethod(methodName, args)
        }
    }

    // BannerAdListener callbacks
    override fun onAdLoaded(baseAd: BaseAd) {
        Log.d(TAG, "onAdLoaded - listenerId: $listenerId")
        sendCallback("onBannerAdLoaded")
    }

    override fun onAdStart(baseAd: BaseAd) {
        Log.d(TAG, "onAdStart - listenerId: $listenerId")
        sendCallback("onBannerAdShowed")
    }

    override fun onAdImpression(baseAd: BaseAd) {
        Log.d(TAG, "onAdImpression - listenerId: $listenerId")
        sendCallback("onBannerAdImpression")
    }

    override fun onAdEnd(baseAd: BaseAd) {
        Log.d(TAG, "onAdEnd - listenerId: $listenerId")
        sendCallback("onBannerAdClosed")
    }

    override fun onAdClicked(baseAd: BaseAd) {
        Log.d(TAG, "onAdClicked - listenerId: $listenerId")
        sendCallback("onBannerAdClicked")
    }

    override fun onAdLeftApplication(baseAd: BaseAd) {
        Log.d(TAG, "onAdLeftApplication - listenerId: $listenerId")
    }

    override fun onAdFailedToLoad(baseAd: BaseAd, adError: VungleError) {
        Log.e(TAG, "onAdFailedToLoad - listenerId: $listenerId, error: ${adError.localizedMessage}")
        sendCallback("onBannerAdLoadFailed", mapOf(
            "error" to mapOf("code" to adError.code, "message" to adError.localizedMessage)
        ))
    }

    override fun onAdFailedToPlay(baseAd: BaseAd, adError: VungleError) {
        Log.e(TAG, "onAdFailedToPlay - listenerId: $listenerId, error: ${adError.localizedMessage}")
        sendCallback("onBannerAdLoadFailed", mapOf(
            "error" to mapOf("code" to adError.code, "message" to adError.localizedMessage)
        ))
    }
}
